finagle-core/src/main/scala/com/twitter/finagle/util/LoadService.scala

```diff
@@ -68,17 +68,22 @@ private object ClassPath {
     ents
   }
 
-  private[finagle] def browseUri(uri: URI, loader: ClassLoader, buf: mutable.Buffer[Info]) {
+  private[finagle] def browseUri(uri: URI, loader: ClassLoader, buf: mutable.Buffer[Info], history: mutable.Set[String] = mutable.Set[String]()) {
     if (uri.getScheme != "file")
       return
     val f = new File(uri)
     if (!(f.exists() && f.canRead))
       return
+    val path = f.getCanonicalPath
```

Comment:
Assume we have a test.jar, which references dot/test.jar by the Class-Path manifest attribute, where dot is a symbolic link to the current directory. In this case, test.jar will be processed repeatedly, because test.jar is not equal to dot/test.jar as File objects, and dot/test.jar is not equals to dot/dot/test.jar, ...

There may never be such a case in real systems, but generally speaking, compare files by canonical path is a safer option.